### PR TITLE
En 7193/end of epoch sc proc

### PR DESCRIPTION
**Pull request by sasurobert**
end of epoch system smart contract processing. Jailing in the system smart contract the low rating nodes, removing them from staking list, adding users from the waiting list to active.

Start a testnet with MaxNumberOfNodesForStake = number of nodes you run from first epoch.
Stake a set of new nodes, they will enter the waiting list
Turn off some of the initial nodes - only a few of them in order to keep the block production going.
After those nodes rating decreases to minimum, they are jailed and replaced with the ones from the waiting list.

### CODE REVIEW

**Comment by nouserboop** [comment]
How to join waiting list please?

File: process/scToProtocol/stakingToPeer.go [hunk 1]
```diff
@@ -227,7 +227,7 @@ func (stp *stakingToPeer) updatePeerState(
 			account.SetTempRating(stp.unJailRating)
 		}
 
-		if !isValidator && account.GetUnStakedEpoch() == core.DefaultUnstakedEpoch {
+		if !isValidator && account.GetUnStakedEpoch() == core.DefaultUnstakedEpoch && stakingData.Staked == true {
```
Comment (iulianpascalau) [inline process/scToProtocol/stakingToPeer.go:230]: Maybe extract in a bool var?

File: process/transaction/shardProcess_test.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ import (
 	"crypto/rand"
 	"encoding/hex"
 	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
```
Comment (iulianpascalau) [inline process/transaction/shardProcess_test.go:8]: unsorted imports

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -999,6 +1008,49 @@ func (r *stakingSC) createWaitingListKey(blsKey []byte) []byte {
 	return []byte(waitingElementPrefix + string(blsKey))
 }
 
+func (r *stakingSC) switchJailedWithWaiting(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1011]: tests?

File: epochStart/metachain/systemSCs.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	marshalizer             marshal.Marshalizer
+	peerAccountsDB          state.AccountsAdapter
+	startRating             uint32
+	validatorInfoCreator    epochStart.ValidatorInfoCreator
+	endOfEpochCallerAddress []byte
+	stakingSCAddress        []byte
+}
+
+// NewSystemSCProcessor creates the end of epoch system smart contract processor
+func NewSystemSCProcessor(args ArgsNewEpochStartSystemSCProcessing) (*systemSCProcessor, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:44]: wanted to have the first review before writing tests.

File: epochStart/metachain/systemSCs.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	vmOutput *vmcommon.VMOutput,
+) error {
+	stakingSCOutput, ok := vmOutput.OutputAccounts[string(s.stakingSCAddress)]
+	if !ok {
+		return epochStart.ErrStakingSCOutputAccountNotFound
+	}
+
+	var activeStorageUpdate *vmcommon.StorageUpdate
+	for _, storageUpdate := range stakingSCOutput.StorageUpdates {
+		if len(storageUpdate.Offset) == len(jailedValidator.PublicKey) &&
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:141]: how does a big bool variable look like ? :dango:

File: integrationTests/mock/epochStartSystemSCStub.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data/state"
+
+type EpochStartSystemSCStub struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/mock/epochStartSystemSCStub.go:5]: done

File: process/mock/epochStartSystemSCStub.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data/state"
+
+type EpochStartSystemSCStub struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/mock/epochStartSystemSCStub.go:5]: done

File: integrationTests/testProcessorNode.go [hunk 8]
```diff
@@ -3,6 +3,7 @@ package integrationTests
 import (
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/vm"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testProcessorNode.go:6]: new goland does not works as expected. done

File: process/common.go [hunk 9]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/common.go:7]: done

File: process/common.go [hunk 10]
```diff
@@ -652,3 +653,42 @@ func IsAllowedToSaveUnderKey(key []byte) bool {
 	trimmedKey := key[:prefixLen]
 	return !bytes.Equal(trimmedKey, []byte(core.ElrondProtectedKeyPrefix))
 }
+
+// SortVMOutputInsideData returns the output accounts as a sorted list
+func SortVMOutputInsideData(vmOutput *vmcommon.VMOutput) []*vmcommon.OutputAccount {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/common.go:658]: in the smart contract process there are tests directly using this function

File: process/transaction/shardProcess_test.go (hunk 2 above)
Comment (sasurobert) [reply process/transaction/shardProcess_test.go:8]: done

File: process/scToProtocol/stakingToPeer.go (hunk 1 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:230]: done

File: epochStart/metachain/systemSCs.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:141]: done

File: vm/systemSmartContracts/staking.go [hunk 11]
```diff
@@ -999,6 +1008,49 @@ func (r *stakingSC) createWaitingListKey(blsKey []byte) []byte {
...
+		return vmcommon.UserError
+	}
+	err = r.moveFirstFromWaitingToStakedIfNeeded(args.Arguments[0])
+	if err != nil {
+		r.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	r.removeFromStakedNodes()
+	registrationData.Staked = false
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1041]: registration data will contain the information related to the waiting or to the jailed node?

File: process/scToProtocol/stakingToPeer.go [hunk 12]
```diff
@@ -227,7 +227,8 @@ func (stp *stakingToPeer) updatePeerState(
 			account.SetTempRating(stp.unJailRating)
 		}
 
-		if !isValidator && account.GetUnStakedEpoch() == core.DefaultUnstakedEpoch && stakingData.Staked == true {
+		isNewValidator := !isValidator && account.GetUnStakedEpoch() == core.DefaultUnstakedEpoch && stakingData.Staked == true
```
Comment (SebastianMarian) [inline process/scToProtocol/stakingToPeer.go:230]: ... && stakingData.Staked is enough

File: epochStart/metachain/systemSCs.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+	peerAccountsDB          state.AccountsAdapter
+	startRating             uint32
+	validatorInfoCreator    epochStart.ValidatorInfoCreator
+	endOfEpochCallerAddress []byte
+	stakingSCAddress        []byte
+}
+
+// NewSystemSCProcessor creates the end of epoch system smart contract processor
+func NewSystemSCProcessor(args ArgsNewEpochStartSystemSCProcessing) (*systemSCProcessor, error) {
+	if check.IfNilReflect(args.SystemVM) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:45]: yep. systemVM interface does not have isInterfacenil

File: process/block/metablock.go [hunk 14]
```diff
@@ -360,22 +365,22 @@ func (mp *metaProcessor) processEpochStartMetaBlock(
...
 		return err
 	}
 
-	err = mp.epochEconomics.VerifyRewardsPerBlock(header, mp.epochRewardsCreator.GetProtocolSustainabilityRewards())
+	err = mp.validatorStatisticsProcessor.ResetValidatorStatisticsAtNewEpoch(allValidatorsInfo)
 	if err != nil {
 		return err
 	}
 
-	err = mp.scToProtocol.UpdateProtocol(body, header.Nonce)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go:378]: in case of epoch start block is not needed.

[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go:378]: there are no smart contract results in the epoch start block - thus this function did nothing.

File: process/common.go [hunk 15]
```diff
@@ -652,3 +653,42 @@ func IsAllowedToSaveUnderKey(key []byte) bool {
...
+	sort.Slice(vmOutput.DeletedAccounts, func(i, j int) bool {
+		return bytes.Compare(vmOutput.DeletedAccounts[i], vmOutput.DeletedAccounts[j]) < 0
+	})
+	sort.Slice(vmOutput.TouchedAccounts, func(i, j int) bool {
+		return bytes.Compare(vmOutput.TouchedAccounts[i], vmOutput.TouchedAccounts[j]) < 0
+	})
+
+	outPutAccounts := make([]*vmcommon.OutputAccount, len(vmOutput.OutputAccounts))
+	i := 0
+	for _, outAcc := range vmOutput.OutputAccounts {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/common.go:668]: output accounts is a map.

File: process/common.go [hunk 16]
```diff
@@ -652,3 +653,42 @@ func IsAllowedToSaveUnderKey(key []byte) bool {
...
+	})
+
+	return outPutAccounts
+}
+
+// GetSortedStorageUpdates returns the storage updates as a sorted list
+func GetSortedStorageUpdates(account *vmcommon.OutputAccount) []*vmcommon.StorageUpdate {
+	storageUpdates := make([]*vmcommon.StorageUpdate, len(account.StorageUpdates))
+	i := 0
+	for _, update := range account.StorageUpdates {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/common.go:684]: storageUpdates is a map

File: process/scToProtocol/stakingToPeer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go]: done

File: vm/systemSmartContracts/staking.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1041]: it already has.

File: process/scToProtocol/stakingToPeer.go (hunk 12 above)
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:230]: done

File: process/block/metablock.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/metablock.go:378]: ok

File: epochStart/metachain/systemSCs.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply epochStart/metachain/systemSCs.go:45]: ok

File: process/common.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/common.go:668]: ok

File: process/common.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/common.go:684]: ok

**Comment by LucianMincu** [comment]
> How to join waiting list please?

Make sure you follow our validator channels. In the following weeks we will deploy the functions to the mainnet and than it will be just a matter of enable-ing the functions for everyone so you'll need to be ready if you want to make sure you'll get one seat in front of everyone. 👍

File: epochStart/metachain/systemSCs_test.go [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/ElrondNetwork/elrond-go/data/blockchain"
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/data/state/factory"
+	"github.com/ElrondNetwork/elrond-go/data/trie"
+	"github.com/ElrondNetwork/elrond-go/epochStart/mock"
+	"github.com/ElrondNetwork/elrond-go/genesis/process/disabled"
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/hashing/sha256"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	economics2 "github.com/ElrondNetwork/elrond-go/process/economics"
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs_test.go:20]: economics2

**Pull request merged**